src/holesPanel/holesWebView.ts

```diff
@@ -0,0 +1,185 @@
+import * as vscode from 'vscode';
+import { EffektHoleInfo } from './effektHoleInfo';
+
+export function generateWebView(
+  holes: EffektHoleInfo[],
+  cssUri: vscode.Uri,
+): string {
+  function escapeHtml(str: string): string {
+    return String(str).replace(
+      /[&<>"']/g,
+      (c) =>
+        ({
+          '&': '&amp;',
+          '<': '&lt;',
+          '>': '&gt;',
+          '"': '&quot;',
+          "'": '&#39;',
+        })[c]!,
+    );
+  }
+
+  function uniqueByName<T extends { name: string }>(arr: T[]): T[] {
+    return Array.from(new Map(arr.map((x) => [x.name, x])).values());
+  }
+
+  // Helper for dropdown section
+  function renderExpDropdown({
+    title,
+    count,
+    idx,
+    kind,
+    placeholder,
+    itemsHtml,
+  }: {
+    title: string;
+    count: number;
+    idx: number;
+    kind: string;
+    placeholder: string;
+    itemsHtml: string;
+  }) {
+    return `
+      <div class="exp-dropdown-section">
+        <div class="exp-dropdown-header collapsed" onclick="toggleDropdown(this)">
+          <span class="exp-dropdown-toggle">&#9660;</span>
+          <span class="exp-dropdown-title">${escapeHtml(title)} (${count})</span>
+        </div>
+        <div class="exp-dropdown-body hidden" id="${kind}-dropdown-body-${idx}">
+          <input class="filter-box" placeholder="${escapeHtml(placeholder)}" oninput="filterDropdownList(event, '${kind}-dropdown-list-${idx}')" />
+          <div class="bindings-list exp-dropdown-list" id="${kind}-dropdown-list-${idx}">
+            ${itemsHtml}
+          </div>
+        </div>
+      </div>
+    `;
+  }
+
+  return /*html*/ `
+    <!DOCTYPE html>
+    <html lang="en">
+    <head>
+      <meta charset="UTF-8" />
+      <title>Effekt Holes</title>
+      <meta name="viewport" content="width=device-width, initial-scale=1.0">
+      <link href="${cssUri}" rel="stylesheet">
+    </head>
+    <body>
+      <div class="container">
+        <h1>Effekt Holes</h1>
+         ${
+           holes.length === 0
+             ? /*html*/ `<div class="empty">
+    There are no holes in this file.<br/><br/>
+    <span class="desc">
+      Holes are placeholders for missing code.
+      You can create a hole using the <code>&lt;&gt;</code> syntax.
+      For example, you can write a definition without a right-hand side as follows:
+      <pre>def foo = <></pre>
+      You can start to fill in a hole using <code>&lt;{ x }&gt;</code>, where <code>x</code> can be a statement or expression.
+      For example, you can write:
+      <pre>def foo = <{ println("foo"); 42 }></pre>
+      For all holes in the current file, this panel will show information about the types and terms in scope.
```

Comment:
Nitpick: I tried quickly restructuring the default text a bit to follow a chain of thought; double newline is my expectation of a line break.
Also, should this be a `span`, a `div`, or a `p`? IIRC `span`s are for just inline stuff?
```suggestion
      This panel shows information about the types and terms in scope for each typed hole.
      Holes are placeholders for missing code used for type-driven development.

      You can create a hole using the <code>&lt;&gt;</code> syntax.
      For example, you can write a definition without a right-hand side as follows:
      <pre>def foo() = &lt;&gt;</pre>
 
      Use <code>&lt;{ x }&gt;</code> in order to fill in a hole with a statement or expression <code>x</code>, for example:
      <pre>def foo() = &lt;{ println("foo"); 42 }&gt;</pre>
```